// autoswap-sdk/dist/utils/ekubo-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TOKEN_ADDRESSES = exports.poolCache = void 0;
exports.fetchEkuboPools = fetchEkuboPools;
exports.fetchPoolsForPair = fetchPoolsForPair;
exports.fetchAndAnalyzePools = fetchAndAnalyzePools;
exports.fetchEkuboPoolsCached = fetchEkuboPoolsCached;
exports.fetchPoolsForPairCached = fetchPoolsForPairCached;
exports.getTokenSymbol = getTokenSymbol;
exports.formatPoolInfo = formatPoolInfo;
/**
 * Default configuration for Ekubo API
 */
const DEFAULT_CONFIG = {
    baseUrl: 'https://mainnet-api.ekubo.org',
    timeout: 10000, // 10 seconds
};
/**
 * Fetch all pools from Ekubo API
 * @param config Optional configuration
 * @returns Array of pool data
 */
async function fetchEkuboPools(config) {
    const finalConfig = { ...DEFAULT_CONFIG, ...config };
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), finalConfig.timeout);
        const response = await fetch(`${finalConfig.baseUrl}/pools`, {
            signal: controller.signal,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
        });
        clearTimeout(timeoutId);
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        const pools = (await response.json());
        return pools;
    }
    catch (error) {
        if (error instanceof Error) {
            if (error.name === 'AbortError') {
                throw new Error('Request timeout - API took too long to respond');
            }
            throw new Error(`Failed to fetch pools: ${error.message}`);
        }
        throw new Error('Failed to fetch pools: Unknown error');
    }
}
/**
 * Fetch pools for a specific token pair
 * @param token0 First token address
 * @param token1 Second token address
 * @param config Optional configuration
 * @returns Array of pools for the specified token pair
 */
async function fetchPoolsForPair(token0, token1, config) {
    const allPools = await fetchEkuboPools(config);
    return allPools.filter((pool) => (pool.token0 === token0 && pool.token1 === token1) ||
        (pool.token0 === token1 && pool.token1 === token0));
}
/**
 * Fetch and analyze pools with fee information
 * @param config Optional configuration
 * @returns Object containing pools and analysis
 */
async function fetchAndAnalyzePools(config) {
    const pools = await fetchEkuboPools(config);
    return {
        pools,
        totalPools: pools.length,
        poolsWithZeroFee: pools.filter((pool) => pool.fee === '0x0' || pool.fee === '0').length,
        uniqueTokenPairs: new Set(pools.map((pool) => [pool.token0, pool.token1].sort().join('-'))).size,
    };
}
/**
 * Cache for pool data to avoid repeated API calls
 */
class PoolCache {
    constructor(ttl = 5 * 60 * 1000) {
        this.cache = new Map();
        // Default 5 minutes
        this.ttl = ttl;
    }
    /**
     * Get cached pools or fetch from API
     * @param key Cache key
     * @param fetcher Function to fetch data if not cached
     * @returns Pool data
     */
    async get(key, fetcher) {
        const cached = this.cache.get(key);
        const now = Date.now();
        if (cached && now - cached.timestamp < this.ttl) {
            return cached.data;
        }
        const data = await fetcher();
        this.cache.set(key, { data, timestamp: now });
        return data;
    }
    /**
     * Clear the cache
     */
    clear() {
        this.cache.clear();
    }
    /**
     * Get cache statistics
     */
    getStats() {
        return {
            size: this.cache.size,
            keys: Array.from(this.cache.keys()),
        };
    }
}
// Global cache instance
exports.poolCache = new PoolCache();
/**
 * Fetch pools with caching
 * @param config Optional configuration
 * @returns Array of pool data
 */
async function fetchEkuboPoolsCached(config) {
    return exports.poolCache.get('all-pools', () => fetchEkuboPools(config));
}
/**
 * Fetch pools for a specific token pair with caching
 * @param token0 First token address
 * @param token1 Second token address
 * @param config Optional configuration
 * @returns Array of pools for the specified token pair
 */
async function fetchPoolsForPairCached(token0, token1, config) {
    const cacheKey = `pair-${[token0, token1].sort().join('-')}`;
    return exports.poolCache.get(cacheKey, () => fetchPoolsForPair(token0, token1, config));
}
/**
 * Token address mappings for common tokens
 */
exports.TOKEN_ADDRESSES = {
    ETH: '0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7',
    USDC: '0x53c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8',
    USDT: '0xda114221cb83fa859dbdb4c44beeaa0bb37c7537ad5ae66fe5e0efd20e6eb3',
    STRK: '0x4718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d',
    WBTC: '0x3fe2b97c1fd336e750087d68b9b867997fd64a2661ff3ca5a7c771641e8e7ac',
};
/**
 * Get token symbol from address
 * @param address Token address
 * @returns Token symbol or null if not found
 */
function getTokenSymbol(address) {
    for (const [symbol, tokenAddress] of Object.entries(exports.TOKEN_ADDRESSES)) {
        if (tokenAddress === address) {
            return symbol;
        }
    }
    return null;
}
/**
 * Format pool information for display
 * @param pool Pool data
 * @returns Formatted pool information
 */
function formatPoolInfo(pool) {
    const token0Symbol = getTokenSymbol(pool.token0) || 'Unknown';
    const token1Symbol = getTokenSymbol(pool.token1) || 'Unknown';
    return `${token0Symbol}/${token1Symbol} - Fee: ${pool.fee} - Key: ${pool.key_hash}`;
}
