// autoswap-sdk/dist/utils/ekubo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchEkuboTopPools = fetchEkuboTopPools;
exports.convertEkuboTopPoolToConfig = convertEkuboTopPoolToConfig;
exports.getTopPoolAsConfig = getTopPoolAsConfig;
/**
 * Fetches pools from the Ekubo API for a specific token pair
 * @param token0 First token address
 * @param token1 Second token address
 * @param options Optional parameters for fetching pools
 * @returns Promise with pool data from the API
 */
async function fetchEkuboTopPools(token0, token1) {
    const baseUrl = 'https://mainnet-api.ekubo.org/pair';
    const url = new URL(`${baseUrl}/${token0}/${token1}/pools`);
    console.log(url);
    try {
        const response = await fetch(url.toString(), {
            method: 'GET',
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
            },
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        const data = await response.json();
        return data;
    }
    catch (error) {
        throw new Error(`Failed to fetch pools from Ekubo API: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
}
/**
 * Converts Ekubo pool data to the internal PoolConfig format
 * @param ekuboPool Pool data from Ekubo API
 * @returns PoolConfig object
 */
function convertEkuboTopPoolToConfig(token0, token1, ekuboPool) {
    return {
        token0,
        token1,
        fee: ekuboPool.fee,
        tick_spacing: ekuboPool.tick_spacing,
        extension: ekuboPool.extension,
        sqrt_ratio_limit: '18446748437148339061',
    };
}
/**
 * Fetches and converts pools to PoolConfig format
 * @param token0 First token address
 * @param token1 Second token address
 * @returns Promise with array of PoolConfig objects
 */
async function getTopPoolAsConfig(token0, token1) {
    const response = await fetchEkuboTopPools(token0, token1);
    return convertEkuboTopPoolToConfig(token0, token1, response.topPools[0]);
}
