// autoswap-sdk/dist/utils/fee-calculator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hexFeeToBigInt = hexFeeToBigInt;
exports.sqrtRatioToBigInt = sqrtRatioToBigInt;
exports.isValidPool = isValidPool;
exports.findLowestFeePool = findLowestFeePool;
exports.findLowestFeePoolForPair = findLowestFeePoolForPair;
exports.getFeeBigInt = getFeeBigInt;
exports.sortPoolsByFee = sortPoolsByFee;
exports.getPoolsForPairSortedByFee = getPoolsForPairSortedByFee;
/**
 * Convert hex fee to BigInt for comparison
 * @param feeHex Hex string representing the fee
 * @returns BigInt representation of the fee, or null if fee is zero
 */
function hexFeeToBigInt(feeHex) {
    const cleanHex = feeHex.startsWith('0x') ? feeHex.slice(2) : feeHex;
    const feeBigInt = BigInt('0x' + cleanHex);
    return feeBigInt;
}
/**
 * Convert sqrt ratio to BigInt for comparison
 * @param sqrtRatioHex Hex string representing the sqrt ratio
 * @returns BigInt representation of the sqrt ratio
 */
function sqrtRatioToBigInt(sqrtRatioHex) {
    const cleanHex = sqrtRatioHex.startsWith('0x')
        ? sqrtRatioHex.slice(2)
        : sqrtRatioHex;
    return BigInt('0x' + cleanHex);
}
/**
 * Check if a pool is valid (non-zero fee, sufficient tick spacing, and sufficient sqrt ratio)
 * @param pool Pool configuration
 * @returns True if pool is valid, false otherwise
 */
function isValidPool(pool) {
    // Check for non-zero fee
    const feeBigInt = hexFeeToBigInt(pool.fee);
    if (feeBigInt === 0n) {
        return false;
    }
    // Check for sufficient tick spacing (>= 1000)
    if (pool.tick_spacing < 1000) {
        return false;
    }
    // Check for sufficient sqrt ratio (>= 18446748437148339061)
    const sqrtRatioBigInt = sqrtRatioToBigInt(pool.sqrt_ratio);
    const minSqrtRatio = 18446748437148339061n;
    if (sqrtRatioBigInt < minSqrtRatio) {
        return false;
    }
    return true;
}
/**
 * Find the pool with the highest liquidity among pools with the lowest fee
 * @param pools Array of pool configurations
 * @returns The pool with highest liquidity among lowest fee pools, or null if no valid pools provided
 */
function findLowestFeePool(pools) {
    if (!pools || pools.length === 0) {
        return null;
    }
    let validPools = [];
    // First, filter out invalid pools
    for (const pool of pools) {
        if (isValidPool(pool)) {
            validPools.push(pool);
        }
    }
    if (validPools.length === 0) {
        return null;
    }
    // Find the lowest fee among valid pools
    let lowestFee = null;
    for (const pool of validPools) {
        const currentFee = hexFeeToBigInt(pool.fee);
        if (lowestFee === null || currentFee < lowestFee) {
            lowestFee = currentFee;
        }
    }
    // Filter pools to only those with the lowest fee
    const lowestFeePools = validPools.filter((pool) => {
        const currentFee = hexFeeToBigInt(pool.fee);
        return currentFee === lowestFee;
    });
    // Among the lowest fee pools, find the one with highest liquidity
    let highestLiquidityPool = lowestFeePools[0];
    let highestLiquidity = BigInt(lowestFeePools[0].liquidity);
    for (let i = 1; i < lowestFeePools.length; i++) {
        const currentLiquidity = BigInt(lowestFeePools[i].liquidity);
        if (currentLiquidity > highestLiquidity) {
            highestLiquidity = currentLiquidity;
            highestLiquidityPool = lowestFeePools[i];
        }
    }
    return highestLiquidityPool;
}
/**
 * Find pools for a specific token pair and return the one with lowest fee
 * @param pools Array of pool configurations
 * @param token0 First token address
 * @param token1 Second token address
 * @returns The pool with the lowest fee for the specified token pair, or null if not found
 */
function findLowestFeePoolForPair(pools, token0, token1) {
    const matchingPools = pools.filter((pool) => (pool.token0 === token0 && pool.token1 === token1) ||
        (pool.token0 === token1 && pool.token1 === token0));
    if (matchingPools.length === 0) {
        return null;
    }
    return findLowestFeePool(matchingPools);
}
/**
 * Get fee as BigInt from hex fee value
 * @param feeHex Hex string representing the fee
 * @returns Fee as BigInt, or null if fee is zero
 */
function getFeeBigInt(feeHex) {
    return hexFeeToBigInt(feeHex);
}
/**
 * Sort pools by fee (lowest to highest), excluding invalid pools
 * @param pools Array of pool configurations
 * @returns Sorted array of pools with valid fees and tick spacing
 */
function sortPoolsByFee(pools) {
    return [...pools]
        .filter((pool) => isValidPool(pool))
        .sort((a, b) => {
        const feeA = hexFeeToBigInt(a.fee); // Safe to use ! here since we filtered
        const feeB = hexFeeToBigInt(b.fee);
        return Number(feeA - feeB);
    });
}
/**
 * Get all pools for a specific token pair, sorted by fee
 * @param pools Array of pool configurations
 * @param token0 First token address
 * @param token1 Second token address
 * @returns Sorted array of pools for the specified token pair
 */
function getPoolsForPairSortedByFee(pools, token0, token1) {
    const matchingPools = pools.filter((pool) => (pool.token0 === token0 && pool.token1 === token1) ||
        (pool.token0 === token1 && pool.token1 === token0));
    return sortPoolsByFee(matchingPools);
}
