// autoswap-sdk/dist/autoswappr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoSwappr = void 0;
const starknet_1 = require("starknet");
const types_1 = require("./types");
const autoswappr_abi_1 = require("./contracts/autoswappr-abi");
const ekubo_1 = require("./utils/ekubo");
const pools_1 = require("./constants/pools");
/**
 * AutoSwappr SDK for interacting with the AutoSwappr Contract
 */
class AutoSwappr {
    constructor(config) {
        this.config = config;
        this.provider = new starknet_1.RpcProvider({ nodeUrl: config.rpcUrl });
        this.account = new starknet_1.Account(this.provider, config.accountAddress, config.privateKey);
        this.autoswapprContract = new starknet_1.Contract(autoswappr_abi_1.AUTOSWAPPR_ABI, config.contractAddress, this.account);
    }
    /**
     * Create swap data for ekubo_manual_swap
     * @param tokenIn Input token address
     * @param tokenOut Output token address
     * @param options Swap options
     * @returns SwapData object
     */
    async createSwapData(tokenIn, tokenOut, options) {
        // Get pool configuration
        const poolConfig = await (0, ekubo_1.getTopPoolAsConfig)(tokenIn, tokenOut);
        if (!poolConfig) {
            throw new Error(types_1.AutoSwapprError.INVALID_POOL_CONFIG);
        }
        // Determine if input token is token1
        const isToken1 = options.isToken1 ?? tokenIn === poolConfig.token1;
        const amountInDecimals = Number(options.amount) * 10 ** pools_1.TOKEN_INFO[tokenIn].decimals;
        // Create swap parameters
        const swapParams = {
            amount: {
                mag: starknet_1.cairo.uint256(amountInDecimals.toString()),
                sign: false,
            },
            sqrt_ratio_limit: options.sqrtRatioLimit || poolConfig.sqrt_ratio_limit,
            is_token1: isToken1,
            skip_ahead: options.skipAhead || 0,
        };
        // Create pool key
        const poolKey = {
            token0: poolConfig.token0,
            token1: poolConfig.token1,
            fee: poolConfig.fee,
            tick_spacing: poolConfig.tick_spacing,
            extension: poolConfig.extension,
        };
        return {
            params: swapParams,
            pool_key: poolKey,
            caller: this.account.address,
        };
    }
    /**
     * Execute swap
     * @param tokenIn Input token address
     * @param tokenOut Output token address
     * @param options Swap options
     * @returns Swap result
     */
    async executeSwap(tokenIn, tokenOut, options) {
        try {
            // Validate inputs
            if (!options.amount || options.amount === '0') {
                throw new Error(types_1.AutoSwapprError.ZERO_AMOUNT);
            }
            // Create swap data
            const swapData = await this.createSwapData(tokenIn, tokenOut, options);
            const amountInDecimals = Number(options.amount) * 10 ** pools_1.TOKEN_INFO[tokenIn].decimals;
            const approveCall = {
                contractAddress: tokenIn,
                entrypoint: 'approve',
                calldata: starknet_1.CallData.compile({
                    spender: this.config.contractAddress,
                    amount: starknet_1.cairo.uint256(amountInDecimals.toString()),
                }),
            };
            const swapCall = {
                contractAddress: this.config.contractAddress,
                entrypoint: 'ekubo_manual_swap',
                calldata: starknet_1.CallData.compile({
                    swapData,
                }),
            };
            const result = await this.account.execute([approveCall, swapCall], undefined, {
                maxFee: '100000000000000',
            });
            return { result };
        }
        catch (error) {
            console.error('Swap failed:', error);
            throw error;
        }
    }
}
exports.AutoSwappr = AutoSwappr;
